core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -694,6 +698,87 @@ public void run() {
         }).drain();
   }
 
+  /**
+   * A registry that prevents channel shutdown from killing existing retry attempts that are in
+   * backoff.
+   */
+  // TODO(zdapeng): add test coverage for shutdown during retry backoff once retry backoff is
+  //                implemented.
+  private final class UncommittedRetriableStreamsRegistry {
+    final Object lock = new Object();
+
+    @GuardedBy("lock")
```

Comment:
This means we would acquire a lock for each new retry-able stream.
The non-retry-able stream has a fast path that is lock-free. Maybe we should spend some time to see if we can make the retry-able path lock-free. Please add a TODO.

Reply:
Added TODO.